### ISSUE TITLE
ARM Linux build (Raspberry Pi OS)

### ISSUE THREAD

**Issue by EugeneUvin** [jacquetc/skribisto#103]
When trying to follow instructions on ARM Linux (Raspberry Pi OS) I've ran into the issues below. Were there previous attempts for building for ARM, and are there any plans to pack it for ARM?

```
uvin@ilotomo:~/skribisto $ flatpak install flathub org.kde.Sdk//5.15

Note that the directories 

'/var/lib/flatpak/exports/share'
'/home/uvin/.local/share/flatpak/exports/share'

are not in the search path set by the XDG_DATA_DIRS environment variable, so
applications installed by Flatpak may not appear on your desktop until the
session is restarted.

Looking for matches…
ошибка: Nothing matches org.kde.Sdk in remote flathub
uvin@ilotomo:~/skribisto $ flatpak install flathub org.kde.Platform//5.15

Note that the directories 

'/var/lib/flatpak/exports/share'
'/home/uvin/.local/share/flatpak/exports/share'

are not in the search path set by the XDG_DATA_DIRS environment variable, so
applications installed by Flatpak may not appear on your desktop until the
session is restarted.

Looking for matches…
ошибка: Nothing matches org.kde.Platform in remote flathub

```

**Comment by jacquetc** [comment]
Hello,

Thank you for trying Skribisto on ARM. We will try to make it work together :-)

No, I didn't try to create an ARM flatpak. I thought it would be flawless.= thanks to Flatpak.
If I read correctly [this](https://github.com/KDE/flatpak-kde-runtime/blob/769cfd36644ecc598aeeb6f58015af8f1f7994a7/org.kde.Sdk.json#L106) , there is a build of KDE runtime for ARM.

Just to be sure : Did you add the flathub remote ?
flatpak remote-add --if-not-exists flathub https://dl.flathub.org/repo/flathub.flatpakrepo

Maybe there is no kde runtime for arm ?

To circumvent that, I only see as fastest solution : 
- install QT Creator, 
-  Qt5 dépendencies (dev)
-- search for a qt5-default or qt5-devel package ?
-  quazip (dev), hunspell (dev) and maybe zlib (dev).
- clone the repository
- switch to "develop" branch (yes, lot of improvements there)
- open the CMakeLists.txt with Qt Creator
-  build it (run it if you want)

I have an arm linux VM on my Mac M1, I'll try it when I can.

**Comment by jacquetc** [comment]
It works for me. Do you have Arm 32bits or Arm 64 bits ?

![image](https://user-images.githubusercontent.com/11695581/108088887-42436980-7079-11eb-8b07-10b81a91b63c.png)

**Comment by EugeneUvin** [comment]
Hi Cyril,

I have 32 bit Raspberry Pi OS I use as a home PC. I'll try one more time
later, did not get to it yet.

Thank you for looking into this, I'm curious to see it built! I have a
little experience so it will take some time :-)

вт, 16 февр. 2021 г., 18:06 Cyril Jacquet <notifications@github.com>:

**Comment by jacquetc** [comment]
Hello,

I don't have any ARM 32bit and I can't emulate it in QEmu for an obscure reason.
Anyway, please try this first step and report the result. Other steps will follow if it succeeds.

``` bash
sudo apt update
sudo apt dist-upgrade
sudo apt install  libfontconfig1-dev libfreetype6-dev libx11-dev libx11-xcb-dev libxext-dev libxfixes-dev libxi-dev libxrender-dev libxcb1-dev libxcb-glx0-dev libxcb-keysyms1-dev libxcb-image0-dev libxcb-shm0-dev  libxcb-icccm4-dev libxcb-sync-dev libxcb-xfixes0-dev libxcb-shape0-dev libxcb-randr0-dev libxcb-render-util0-dev libxcb-util-dev libxcb-xinerama0-dev  libxkbcommon-dev libxkbcommon-x11-dev llvm libatspi2.0-dev libcairo-script-interpreter2 libjpeg-turbo8-dev libdrm-freedreno1 libice-dev libepoxy-dev libegl1-mesa-dev libjpeg-dev libcairo2-dev libgles-dev libcups2-dev libcupsimage2-dev libharfbuzz-gobject0 wayland-protocols libsm-dev libtiff-dev libopengl0 x11proto-xinerama-dev libdrm-etnaviv1 libxrandr-dev libfribidi-dev libgles1 libxinerama-dev libcupsfilters-dev pango1.0-tools libpango1.0-dev libwayland-bin libharfbuzz-dev libdrm-tegra0 x11proto-randr-dev libwayland-dev libxcursor-dev libjbig-dev libgtk-3-dev libglvnd-dev libxft-dev libpulse-dev libmd4c-dev libgl1-mesa-dev libgraphite2-dev libgdk-pixbuf2.0-dev libclang-11-dev libxcomposite-dev libicu-dev libjpeg8-dev libpixman-1-dev libthai-dev libxdamage-dev icu-devtools libopengl-dev libdrm-dev libatk1.0-dev libatk-bridge2.0-dev liblzma-dev libtiffxx5 libdatrie-dev  libpcre16-3 python3-distutils libpcre3-dev libxtst-dev x11proto-record-dev libatspi2.0-dev libselinux1-dev libpcre32-3 libmount-dev libglib2.0-dev-bin libpcre2-dev libdbus-1-dev python3-lib2to3 libglib2.0-dev libsepol1-dev libpcre2-posix2 libpcrecpp0v5 libblkid-dev  


mkdir ~/Devel && cd ~/Devel
git clone git://code.qt.io/qt/qt5.git
cd qt5
git checkout 5.15.2
perl init-repository
cd ~/Devel && mkdir qt5-build && cd qt5-build
../qt5/configure -opensource -shared -nomake examples -nomake tests -confirm-license -c++std 14 -recheck-all -skip webengine -skip lottie -skip gamepad -skip quick3d -skip 3d -skip location -release -prefix ~/Devel/Qt
make -j2
make install

```

**Comment by EugeneUvin** [comment]
Hi Cyril,

RPi OS is 32 bit Debian ARM installed to 64 bit machine, so I assume it can
be emulated in 64 bit VM.

When trying to install the big list of packages following were not found:
E: Невозможно найти пакет* libxcb-util-dev*
E: Для пакета «*libjpeg-turbo8-dev*» не найден кандидат на установку
E: Невозможно найти пакет* libmd4c-dev*
E: Невозможно найти пакет *libclang-11-dev*
E: Невозможно найти пакет *libpcre2-posix2*

When I've unlisted them some conflicts appeared (citing output in Russian,
basically it requires packages on the right):
Следующие пакеты имеют неудовлетворённые зависимости:
 libjpeg-dev : Конфликтует: libjpeg8-dev но 8d1-2 должен быть установлен
 libjpeg62-turbo-dev : Конфликтует: libjpeg8-dev но 8d1-2 должен быть
установлен

Then, I've tried to build qt5 and ran of the space on SD card which is used
as main storage on my Pi.

I'll need to think how to work this around...

Thanks.
Eugene

ср, 17 февр. 2021 г. в 14:17, Cyril Jacquet <notifications@github.com>:

**Comment by jacquetc** [comment]
Hello, 

I'm using an Ubuntu Arm VM (64bit). Which Raspberry are you using ?
Search the exact names of your missing packages manually with `sudo apt search libxcb-util` Same with libclang, etc...

Clean up with `rm -fr ~/Devel/qt5-build`

Insert and mount a USB key
Type :
```
 cd [your USB key path] && mkdir qt5-build && cd qt5-build
~/Devel/qt5/configure -opensource -shared -nomake examples -nomake tests -confirm-license -c++std 14 -recheck-all -skip webengine -skip lottie -skip gamepad -skip quick3d -skip 3d -skip location -release -prefix ~/Devel/Qt
make -j2
make install
```

**Comment by EugeneUvin** [comment]
Hi,

I don't think Debian ARM has those packages at all, I've searched for them. However, I'll try to do that with my friend's USB. Those USB sticks I have are way too old to have enough memory on them.

I use Raspberry Pi 4 8 GB, Raspberry Pi OS official build - i.e. Debian on ARM 32 bit. They use 32 bit OS for consistency through all devices they support - oldest ones being 32 bit.

Thank you for investing time into this.

**Comment by jacquetc** [comment]
Hello,

For another project, I have a Raspberry 4 with "Raspbian" on it. The command "name -m" gives me armv7l, so 32 bit too. Ultimately, I wanted to port Skribisto on it. I'll test a few things on my side.

**Comment by jacquetc** [comment]
I found out there where kde 5.12 and 5.14, but no 5.15. Do a `flatpack search ode`. I go looking why there is no 5.15

**Comment by EugeneUvin** [comment]
You're right, there is no 5.15:

```
$ flatpak search kde|grep 5.15
$ flatpak search kde|grep 5.14
KDE Software Development Kit - Tools and headers for developing applications using the KDE application platform	org.kde.Sdk		5.14	flathub
KDE Application Platform - Shared libraries used by KDE applications	org.kde.Platform		5.14	flathub
QGnomePlatform-decoration - Qt Platform Theme aimed to accommodate GNOME settings	org.kde.WaylandDecoration.QGnomePlatform-decoration		5.14	flathub
QtSNI - A Qt platformtheme plugin that offers a flatpak-compatible SNI tray icon	org.kde.PlatformTheme.QtSNI		5.14	flathub
QGnomePlatform - Qt Platform Theme aimed to accommodate GNOME settings	org.kde.PlatformTheme.QGnomePlatform		5.14	flathub
Maliit plugin for SailfishOS - Keyboard plugin used in Sailfish OS for Flatpak applications	org.kde.PlatformInputContexts.MaliitSailfishOS		5.14	flathub
HighContrast Theme - HighContrast widget theme matching the GNOME highcontrast theme	org.kde.KStyle.HighContrast		5.14	flathub
Adwaita theme - Adwaita widget theme matching the GNOME adwaita theme	org.kde.KStyle.Adwaita		5.14	flathub
```

**Comment by jacquetc** [comment]
Let me try to built it against Qt 5.14.2. I have to change a few lines of code to do that.

**Comment by jacquetc** [comment]
Don't wait for a kde 5.15 on arm 32bit : [see this forum](https://invent.kde.org/plasma-mobile/plasma-angelfish/-/issues/34). Freedesktop (the basis for kde runtime) seems to have  dropped the support.

I sadly can't backport to 5.14 : too much lines unique to 5.15.

I'm exploring other means to deploy on arm (snapd ? cross-compilation ? static ? ...)

**Comment by jacquetc** [comment]
Do you know that there is a 64 bit version of Raspberry Pi OS (the new name of Raspbian) ? It's still beta, but it runs well on my Pi 4. I installed it after I broke my system today. There is also an Ubuntu Desktop 20.04 64 bit available.

**Comment by EugeneUvin** [comment]
64 bit version was behaving unstable with Chromium video - and my main use
case is basically using browsers on it so far. I'll see if things improved
since, thank you.

32 bit is representative as it is what they propose to install by default
to RPi so far. Hopefully 64 bit will go out of beta soon and gets this
burden off developers shoulders :-)

пт, 19 февр. 2021 г. в 20:44, Cyril Jacquet <notifications@github.com>:

**Comment by EugeneUvin** [comment]
Hi @jacquetc, I've booted up 64 bit RPiOS and built the app - not sure if it looks right, but it definitely works. I'll think if I'd want to switch to 64 bit now, but I am glad that I could run your app on RPi!
![2021-02-20-221239_1920x1080_scrot](https://user-images.githubusercontent.com/9447963/108607430-2d443e80-73c9-11eb-9dd6-4f3bac03bf5c.png)

**Comment by jacquetc** [comment]
Hello Eugene,

Great work ! Glad that you could have tested. I hope that you are not disappointed : all features aren't there yet. Moreover, I did all kinds of deep changes and not all my work is migrated on master yet.

Try to test Ubuntu 64 bit, I didn't read any big issue with videos. I'll also test it in the week.

Yes, this is Skribisto that I see on your screen. I'm adding features every week in develop branch, so feel free to ask questions or give your thoughts :-)

**Comment by jacquetc** [comment]
For you to know : I did deep changes in the code in the last weeks. It means some feature which were already implemented aren't yet migrated to the most current version. 

Changes were :
- setting a plugin system to add whatever type of document we would need in the future. Right now, I have Project, Folder, Text. Soon, I'll add Section. 
- merge the old Sheet tree and Note tree in only one generic tree, thus allowing the plugin system
- add Folders logic where there were none before
- add window management
- add the split views management
- new theme logic
- new Navigation

Features yet to be migrated from the older version of Skribisto:
- modify theme page
- adapt import / export / print to the plugin system
- rethink the "main menu" page and decide the best way  to display it
- fix the accessibilty

**Comment by jacquetc** [comment]
If you want to check my latest work, you only have to change the last line of `package/flatpak/eu.skribisto.skribisto.yml`. from `master` to `develop` and redo the same build steps for flatpack

**Comment by EugeneUvin** [comment]
Hi @jacquetc , I've tried replasing `master` with `develop` in yml and it won't build:

```
========================================================================
Building module skribisto in /home/pi/Devel/.flatpak-builder/build/skribisto-6
========================================================================
Switched to branch 'develop'
-- Found Git: /usr/bin/git (found version "2.30.0") 
CMake Error at cmake/3rdparty/MunkeyVersionFromGit.cmake:65 (message):
  [MunkeiVersionFromGit] Failed to execute Git: fatal: No tags can describe
  '75cf99c496be965e2210f8610387095a8df4ec01'.

  Try --always, or create some tags.
Call Stack (most recent call first):
  CMakeLists.txt:19 (version_from_git)
```

**Comment by jacquetc** [comment]
Sorry, I wrote a mistake. Put back the "master" in this file. You only have to switch git to develop (`git checkout develop`). Then : 
`cd ~/Devel && flatpak-builder --user --repo=local-repo build-dir skribisto/package/flatpak/local/eu.skribisto.skribisto.yml --force-clean && flatpak install local-repo eu.skribisto.skribisto -y --reinstall`

**Comment by jacquetc** [comment]
And don't forget to do `git pull` to update git

**Comment by EugeneUvin** [comment]
Conclusion: 64 bit build is possible, 32 bit build is not possible - and maybe not required in the near future, as RPiOS default is 32 bit only for the purpose of backwards compatibility that will end in next few years as support for old boards ends.
